core/commands/corenet.go

```diff
@@ -0,0 +1,520 @@
+			return
+		}
+
+		switch lnet {
+		case "tcp", "tcp4", "tcp6":
+			listener, err := manet.Listen(bindAddr)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				if err := remote.Close(); err != nil {
+					res.SetError(err, cmds.ErrNormal)
+				}
+				return
+			}
+
+			app.Address = listener.Multiaddr()
+			app.Closer = listener
+			app.Running = true
+
+			go doAccept(n, &app, remote, listener)
+
+		default:
+			res.SetError(errors.New("unsupported protocol: "+lnet), cmds.ErrNormal)
+			return
+		}
+
+		output := CorenetAppInfoOutput{
+			Protocol: app.Protocol,
+			Address:  app.Address.String(),
+		}
+
+		res.SetOutput(&output)
+	},
+}
+
+func doAccept(n *core.IpfsNode, app *corenet.AppInfo, remote net.Stream, listener manet.Listener) {
+	defer listener.Close()
+
+	local, err := listener.Accept()
+	if err != nil {
+		return
+	}
+
+	stream := corenet.StreamInfo{
+		Protocol: app.Protocol,
+
+		LocalPeer: app.Identity,
+		LocalAddr: app.Address,
+
+		RemotePeer: remote.Conn().RemotePeer(),
+		RemoteAddr: remote.Conn().RemoteMultiaddr(),
+
+		Local:  local,
+		Remote: remote,
+
+		Registry: &n.Corenet.Streams,
+	}
+
+	n.Corenet.Streams.Register(&stream)
+	startStreaming(&stream)
+}
+
+var corenetCloseCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Closes an active stream listener or client.",
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("HandlerID", false, false, "Application listener or client HandlerID"),
+		cmds.StringArg("Protocol", false, false, "Application listener or client HandlerID"),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption("all", "a", "Close all streams and listeners.").Default(false),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		n, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		err = checkEnabled(n)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		if !n.OnlineMode() {
+			res.SetError(errNotOnline, cmds.ErrClient)
+			return
+		}
+
+		closeAll, _, _ := req.Option("all").Bool()
+
+		var proto string
+		var handlerID uint64
+
+		useHandlerID := false
+
+		if !closeAll && len(req.Arguments()) == 0 {
+			res.SetError(errors.New(" handlerID nor stream protocol"), cmds.ErrNormal)
+			return
+
+		} else if !closeAll {
+			handlerID, err = strconv.ParseUint(req.Arguments()[0], 10, 64)
+			if err != nil {
+				proto = "/app/" + req.Arguments()[0]
+
+			} else {
+				useHandlerID = true
+			}
+		}
+
+		if closeAll || useHandlerID {
+			for _, s := range n.Corenet.Streams.Streams {
+				if !closeAll && handlerID != s.HandlerID {
+					continue
+				}
+				s.Close()
+				if !closeAll {
+					break
+				}
+			}
+		}
+
+		if closeAll || !useHandlerID {
+			for _, a := range n.Corenet.Apps.Apps {
+				if !closeAll && a.Protocol != proto {
+					continue
+				}
+				a.Close()
+				if !closeAll {
+					break
+				}
+			}
+		}
+
+		if len(req.Arguments()) != 1 {
+		}
```

Comment:
Yeah, that is why it would be nice to move all logic to its package. It wasn't clear.